tensorboard/webapp/plugins/npmi/util/umap.ts

```diff
@@ -0,0 +1,115 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {UMAP} from 'umap-js';
+import * as util from '../../../../plugins/projector/vz_projector/util';
+
+const UMAP_SAMPLE_SIZE = 20000;
+
+export interface EmbeddingDataPoint {
+  vector: number[];
+  name: string;
+  sequenceIndex?: number;
+  index: number;
+  projections: {
+    [key: string]: number;
+  };
+}
+
+export interface EmbeddingListing {
+  [annotation: string]: EmbeddingDataPoint;
+}
+
+/**
+ * Dataset contains a DataPoints array that should be treated as immutable. This
+ * acts as a working subset of the original data, with cached properties
+ * from computationally expensive operations. Because creating a subset
+ * requires normalizing and shifting the vector space, we make a copy of the
+ * data so we can still always create new subsets based on the original data.
+ */
+export interface EmbeddingDataSet {
+  points: EmbeddingListing;
+  pointKeys: string[];
+  shuffledDataIndices: number[];
+  /**
+   * This keeps a list of all current projections so you can easily test to see
+   * if it's been calculated already.
+   */
+  projections: {
+    [projection: string]: boolean;
+  };
+
+  // UMAP
+  hasUmapRun: boolean;
+  umapRun: number;
+}
+
+export function buildEmbeddingDataSet(
+  points: EmbeddingListing
+): EmbeddingDataSet {
+  const pointKeys = Object.keys(points);
+  return {
+    points,
+    pointKeys,
+    shuffledDataIndices: util.shuffle(util.range(pointKeys.length)),
+    projections: {},
+    hasUmapRun: false,
+    umapRun: 0,
+  };
+}
+
+/** Runs UMAP on the data. */
+export async function projectUmap(
+  embeddingData: EmbeddingDataSet,
+  nNeighbors: number,
+  minDist: number,
+  umapIndices: number[],
+  messageCallback: (message: string) => void,
+  datasetCallback: (dataset: EmbeddingDataSet) => void
+) {
```

Comment:
Specify the return type. Currently, it is set as `Promise<void>`. Might want to remove the datasetCallback and return a `Promise< EmbeddingDataSet>` instead.